lib/features/build_server/build_provider.dart

```diff
@@ -0,0 +1,265 @@
+  }
+
+  /// Handles a new websocket message.
+  Future<void> _handleWebSocketMessage(ServerSession session, dynamic message) async {
+    if (message is ProjectInitRequest) {
+      await _handleProjectInit(session, message);
+      return;
+    }
+
+    final String id = _ensureId(session);
+
+    if (message == reserveBuildRequest) {
+      _projectCapacityQueues[id]!.addLast(session);
+
+      if (_isFirstInQueue(session, id)) {
+        await _activateSession(session, id);
+      }
+    }
+
+    if (message == unsubscribeRequest) {
+      await _activateNextSession(id, removing: session);
+      return;
+    }
+
+    if (!_isFirstInQueue(session, id)) {
+      session.send(isInactiveSessionResponse);
+      return;
+    }
+
+    if (message is ProjectUpdateRequest) {
+      await _handleProjectUpdate(session, message, id);
+    }
+
+    if (message is FileUpdate) {
+      await _handleFileUpdate(session, message, id);
+    }
+
+    if (message == projectBuildRequest) {
+      await _getInterface(id).build();
+    }
+
+    if (message == getBinaryRequest) {
+      await _handleBinaryRequest(session, message, id);
+    }
+  }
+
+  /// Checks if the [session] is the first in the queue for
+  /// the project with the id [id].
+  bool _isFirstInQueue(ServerSession session, String id) {
+    if (_projectCapacityQueues[id]?.isNotEmpty == true) {
+      return identical(_projectCapacityQueues[id]?.first, session);
+    }
+
+    return false;
+  }
+
+  /// Activates the first session in the queue for the project with the id [id].
+  ///
+  /// If [removing] is not null, the session is removed from the queue.
+  Future<void> _activateNextSession(String id, {ServerSession? removing}) async {
+    if (removing != null) {
+      _projectCapacityQueues[id]?.remove(removing);
+    }
+
+    if (_projectCapacityQueues[id]?.isNotEmpty == true) {
+      _activateSession(_projectCapacityQueues[id]!.first, id);
+    }
+  }
+
+  /// Activates the [session] for the project with the id [id].
+  Future<void> _activateSession(ServerSession session, String id) async {
+    _interfaces[id]?.onStateUpdate = session.send;
+    session.send(isActiveSessionResponse);
+  }
+
+  /// Returns the interface with the given [id] or throws an exception.
+  MainInterface _getInterface(String id) {
+    if (_interfaces.containsKey(id)) {
+      return _interfaces[id]!;
+    }
+
+    throw StateError('Interface with id $id has not been created yet.');
+  }
+
+  /// Throws an error if the [session] has no id or returns it.
+  String _ensureId(ServerSession session) {
+    final dynamic id = session.storage['id'];
+
+    if (id is String) {
+      return id;
+    }
+
+    throw StateError('Session has no project id.');
+  }
+
+  /// Handles a new project initialization request.
+  Future<void> _handleProjectInit(ServerSession session, ProjectInitRequest message) async {
+    if (_interfaces.containsKey(message.id)) {
+      return;
```

Comment:
From my understanding, every connecting client will send a `ProjectInitRequest`. I can then try to aquire 'the build lock' with a `reserveBuildRequest`?

Reply:
`ProjectInitRequest` requests to start up a project worker. If you want to connect to an already started project you can skip `ProjectInitRequest`. When sending `reserveBuildRequest` you want access to a specific project. As soon as resources are free, the server responds with a message that grants this access.

Reply:
Correction, you have to send `ProjectInitRequest` anyway, to register for the right project, but only when it's not already started a new worker is created